### PR TITLE
Heartbeat sender factory

### PR DESCRIPTION
**Pull request by sstanculeanu**
## Description of the reasoning behind the pull request (what feature was missing / how the problem was manifesting itself / what was the motive behind the refactoring)
- with 2 types of heartbeat sender, a factory is needed to create the proper one
  
## Proposed Changes
- implement factory

## Testing procedure
- standard system test

### CODE REVIEW

File: heartbeat/sender/commonHeartbeatSender.go [hunk 1]
```diff
@@ -59,3 +66,26 @@ func (chs *commonHeartbeatSender) generateMessageBytes(
 
 	return chs.marshaller.Marshal(msg)
 }
+
+// GetCurrentNodeType will return the current sender type and subtype
+func (chs *commonHeartbeatSender) GetCurrentNodeType() (string, core.P2PPeerSubType, error) {
```
Comment (iulianpascalau) [inline heartbeat/sender/commonHeartbeatSender.go:71]: :+1: less duplicated code

File: heartbeat/sender/heartbeatSenderFactory.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	if err != nil {
+		return false, err
+	}
+
+	keysMap := keysHolder.GetManagedKeysByCurrentNode()
+	isMultikey := len(keysMap) > 0
+
+	_, _, err = nodesCoordinator.GetValidatorWithPublicKey(pkBytes)
+	if err == nil && isMultikey {
+		return false, heartbeat.ErrInvalidConfiguration
```
Comment (iulianpascalau) [inline heartbeat/sender/heartbeatSenderFactory.go:100]: we might return more info here, for easier debugging.
Something like:
```GO
return false, fmt.Errorf("%w, len(keysMap) = %d, isValidator = %v", heartbeat.ErrInvalidConfiguration, len(keysMap) > 0, err == nil)
```

File: heartbeat/sender/heartbeatSenderFactory_test.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+		args := createMockHeartbeatSenderFactoryArgs()
+		args.nodesCoordinator = &shardingMocks.NodesCoordinatorStub{
+			GetValidatorWithPublicKeyCalled: func(publicKey []byte) (validator nodesCoordinator.Validator, shardId uint32, err error) {
+				return nil, 0, nil
+			},
+		}
+		hbSender, err := createHeartbeatSender(args)
+		assert.Nil(t, err)
+		assert.False(t, check.IfNil(hbSender))
+		assert.Equal(t, "*sender.heartbeatSender", fmt.Sprintf("%T", hbSender))
```
Comment (iulianpascalau) [inline heartbeat/sender/heartbeatSenderFactory_test.go:94]: :+1:

File: heartbeat/sender/sender.go [hunk 4]
```diff
@@ -41,7 +42,7 @@ type ArgSender struct {
 
 // sender defines the component which sends authentication and heartbeat messages
 type sender struct {
-	heartbeatSender *heartbeatSender
+	heartbeatSender heartbeatSenderHandler
```
Comment (iulianpascalau) [inline heartbeat/sender/sender.go:45]: :+1: for cleaning up
no way around this, thus.

File: integrationTests/factory/componentsHelper.go [hunk 5]
```diff
@@ -66,9 +66,10 @@ func CreateDefaultConfig() *config.Configs {
 	configs.EpochConfig = epochConfig
 	configs.RoundConfig = roundConfig
 	configs.FlagsConfig = &config.ContextFlagsConfig{
-		WorkingDir: "workingDir",
-		UseLogView: true,
-		Version:    Version,
+		WorkingDir:  "workingDir",
+		UseLogView:  true,
+		BaseVersion: fmt.Sprintf("%s-base", Version),
```
Comment (iulianpascalau) [inline integrationTests/factory/componentsHelper.go:71]: why -base suffix?

**Review by ssd04 — COMMENTED** [review]
No other things from me :+1:

File: integrationTests/factory/componentsHelper.go (hunk 5 above)
Comment (sstanculeanu) [reply integrationTests/factory/componentsHelper.go]: clear now, pushed

**Review by iulianpascalau — APPROVED** [review]

**Review by ssd04 — APPROVED** [review]

**Pull request merged**